### PR TITLE
feat: add epub file reader support

### PR DESCRIPTION
**Pull request by m1911star**
since I have too many epub files, I tried to add epub reader

deps

- https://github.com/gaoxiaoliangz/epub-parser

### CODE REVIEW

**Comment by yisding** [comment]
Thanks for doing this! Is this ready to merge?

**Comment by m1911star** [comment]
> Thanks for doing this! Is this ready to merge?

nope. I need more tests.

File: README.md [hunk 1]
```diff
@@ -97,7 +97,7 @@ export const runtime = "nodejs"; // default
 /** @type {import('next').NextConfig} */
 const nextConfig = {
   experimental: {
-    serverComponentsExternalPackages: ["pdf-parse"], // Puts pdf-parse in actual NodeJS mode with NextJS App Router
+    serverComponentsExternalPackages: ["pdf-parse", "@gxl/epub-parser"], // Puts pdf-parse and @gxl/epub-parser in actual NodeJS mode with NextJS App Router
```
Comment (m1911star) [inline README.md:100]: since I cannot test nextjs app, so I just copy like pdf-parse config here.
Advice needed. thank you

Comment (yisding) [reply README.md:100]: I don't think you necessarily need to put this here, but I'll give it a try with the NextJS configuration and see what happens.

`create-llama` is probably your best bet to test this. @marcusschiesser thoughts?

Comment (marcusschiesser) [reply README.md:100]: @m1911star and @yisding 
yes, in `create-llama` we're just using:
```
serverComponentsExternalPackages: ["llamaindex"]
```
This should work here, too, but explicitly enumerating the problematic dependencies (before just "pdf-parse" and now "@gxl/epub-parser") has the advantage that we know who is the culprit and what we have to do.
This brings us to the question: @m1911star, is there another epub parser package that works with nextjs?
There seem to be alternatives: https://www.npmjs.com/search?q=epub-parser

Comment (m1911star) [reply README.md:100]: I have dived into alternatives of @gxl/epub-parser and did not find the right choice suitable for next app.

Also `@gxl/epub-parser` can not work in next app.

Im working to find another solution. Thanks.

Comment (yisding) [reply README.md:100]: Thanks @m1911star! Appreciate the contribution!

**Comment by EmanuelCampos** [comment]
looks cool! 

bump @m1911star any update here?

File: packages/core/package.json [hunk 2]
```diff
@@ -4,6 +4,12 @@
   "license": "MIT",
   "type": "module",
   "dependencies": {
+    "@anthropic-ai/sdk": "^0.9.1",
+    "@gxl/epub-parser": "2.0.4",
+    "@notionhq/client": "^2.2.13",
+    "@xenova/transformers": "^2.8.0",
+    "crypto-js": "^4.2.0",
+    "js-tiktoken": "^1.0.8",
```
Comment (marcusschiesser) [inline packages/core/package.json:12]: this should only add the epub parser

**Review by marcusschiesser — COMMENTED** [review]
Thanks @m1911star looking good. Please have a look at the comments and also run `pnpx changeset` in root to generate a changeset file

File: packages/core/package.json (hunk 2 above)
Comment (m1911star) [reply packages/core/package.json:12]: Yes, I'm still working on this. Sorry for bothering you.
Will change this pr to draft until it's ready.

**Comment by m1911star** [comment]
let's have a try

File: examples/data/wells.epub
Comment (marcusschiesser) [inline examples/data/wells.epub:1]: where is this example taken from? what license?

File: packages/core/src/readers/EpubReader.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    const sections = book.sections ?? [];
+    const header = `${book.info?.author}\n${book.info?.publisher}\n${book.info?.title}`;
+    const options = this.getOptions();
+    const results = await Promise.all(
+      sections.map((section) => {
+        return new Promise(async (resolve) => {
+          const parsed = await this.parseContent(section.htmlString, options);
+          resolve(parsed);
+        });
+      }),
```
Comment (marcusschiesser) [inline packages/core/src/readers/EpubReader.ts:28]: this doesn't work?

```suggestion
      sections.map(async (section) => {
       return await this.parseContent(section.htmlString, options);
        });
      }),
```

File: packages/core/src/readers/EpubReader.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  getOptions() {
+    return {
+      skipHtmlDecoding: true,
+      stripTogetherWithTheirContents: [
+        "script", // default
+        "style", // default
+        "xml", // default
+        "head", // <-- custom-added
+      ],
+    };
```
Comment (marcusschiesser) [inline packages/core/src/readers/EpubReader.ts:61]: Modifying these options is hard for the user (needs to override the method)

how about passing the options via a constructor to `EpubReader` and using these values above as default? (if the user doesn't pass any options). 

also has `@gxl/epub-parser` a type for the options?

Comment (m1911star) [reply packages/core/src/readers/EpubReader.ts]: the options here is for `string-strip-html`, not parser, so I just make it as the default options and remove getOptions

File: packages/core/src/readers/index.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 export * from "./AssemblyAIReader.js";
 export * from "./CSVReader.js";
 export * from "./DocxReader.js";
+export * from "./EpubReader.js";
```
Comment (himself65) [inline packages/core/src/readers/index.ts:4]: I prefer not to export this to root. And be careful with edge runtime

File: examples/data/wells.epub
Comment (himself65) [inline examples/data/wells.epub:1]: could we download from somewhere instead of leave it here

File: examples/epub.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { EpubReader, VectorStoreIndex } from "llamaindex";
+
+async function main() {
+  // Load PDF
+  const reader = new EpubReader();
+  const documents = await reader.loadData("data/wells.epub");
```
Comment (marcusschiesser) [inline examples/epub.ts:6]: @m1911star you can load the ebub from an URL here instead

**Comment by himself65** [comment]
outdated, closing

**Pull request closed without merging**